src/execution/executor/io.rs

```diff
@@ -0,0 +1,131 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{
+    cell::RefCell,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{Future, FutureExt};
+use tokio::{runtime::Handle, task::JoinHandle};
+
+thread_local! {
+    /// Tokio runtime `Handle` for doing network (I/O) operations, see [`spawn_io`]
+    pub static IO_RUNTIME: RefCell<Option<Handle>> = const { RefCell::new(None) };
+}
+
+/// Registers `handle` as the IO runtime for this thread
+///
+/// See [`spawn_io`]
+pub fn register_io_runtime(handle: Option<Handle>) {
+    IO_RUNTIME.set(handle)
+}
+
+/// Runs `fut` on the runtime registered by [`register_io_runtime`] if any,
+/// otherwise awaits on the current thread
+///
+/// # Panic
+/// Needs a IO runtime [registered](register_io_runtime).
+pub async fn spawn_io<Fut>(fut: Fut) -> Fut::Output
```

Comment:
One thing I don't see in this PR is actually calling `spawn_io` - if this isn't called, then during an ExecutionPlan that is run on the dedicated executor, the IO will be done on the DedicatedExector's threadpool

Here is where influxdb3 calls this:

https://github.com/influxdata/influxdb3_core/blob/6fcbb004232738d55655f32f4ad2385523d10696/iox_query/src/physical_optimizer/cached_parquet_data.rs#L140


I think a better idea is likely to create some sort of ObjectStore wrapper that forwards all network calls to the IO runtime. I'll try and work up some version of that in the DataFusion example.

Reply:
@alamb i understand iOX heavily uses parquet files but based on what i see that means any other types of files would be run on the `DedicatedExecutor` - correct?  I think I will need to come up with a more general purpose solution as the idea with `dft` is to be general purpose / work with all datafusion supported file types (CSV, JSON, Arrow IPC, etc)  - perhaps i will create a wrapping `ObjectStore` where all calls to the underlying `ObjectStore` use `spawn_io`.

Reply:
I am going to add this `ObjectStore` wrapper in a follow on PR.  Then I think this wrapper `ObjectStore` could be used anywhere an `ObjectStore` is needed - like the `ParquetFileReaderFactory`.

Reply:
Awesome -- thank you -- I think your idea is 👌  very good